snuba/consumers/consumer_builder.py

```diff
@@ -119,6 +119,8 @@ def __init__(
             tags={"group": group_id, "storage": storage_key.value},
         )
 
+        configure_metrics(StreamMetricsAdapter(self.metrics))
```

Comment:
I think this is a problem. Here you are configuring the metrics in a place that is not the entry point of the application (while you are also initializing metrics in those entry points). You have to initialize it here because the entry point may not always know it will need to use the streaming library.
This can open the door to initializing metrics inconsistently. You may be changing the metrics object that somebody else somewhere else in the application is using.

The root cause of this is turning metrics into a global mutable variable. 
- it is global as it is a module variable that is shared by everybody who uses the module (the metrics module in the consumer library in this case)
- it is mutable as you have a `configure_metrics` anybody can call changing the global state. Other scenarios in which we use global state (Redis, settings, config, or metrics before this PR) are different because the state is immutable there. It is initialized at the first load of the module and there is no way to mutate it.

Mutable global variable have a lot of downsides which like adding hidden dependencies and coupling. In the case of cross cutting concerns that are not part of your business logic (like loggers and metrics) this is generally acceptable. Though making the variable mutable makes things worse (especially in a mutlithreaded/multi process environment) allowing to make change to shared state. 

\I am not sure the goal of making the api less verbose (and you are right, easier to understand) is worth adding this hidden dependency.

Reply:
I've added code to ensure that the `configure_metrics()` function can only be called once. It can also be never called as metrics is not critical to this library working and the dummy backend is used as a default. I hope this will address some of your concerns.

Of course there is still some global mutable state in the library now though. This may not be ideal but in my opinion this is better than inconsistent and possibly ever changing APIs. Example: right now the user just needs to *know* that ParallelTransform takes metrics but Transform does not. And this is subject to change if we want to record some different stuff in the future. I think an API like this should be avoided at all costs.

Reply:
> Of course there is still some global mutable state in the library now though.

It is, but a variable that can be initialized only once and cannot be changed afterwards is a lot better exactly because you cannot mutate it once others are depending on it.